### PR TITLE
 Prevent storing of metadata that does not have an associated product

### PR DESCRIPTION
**Pull request by jhunzik**
I added a check to ensure that we only store Metadata if we have a product that has the same ID in S3. This PR also clears out Solr between itests to ensure that stored metadata doesn't bleed over into the next test.

### CODE REVIEW

**Comment by jhunzik** [comment]
build now

File: multi-int-store/src/main/java/com/connexta/multiintstore/adaptors/S3StorageAdaptor.java [hunk 1]
```diff
@@ -117,4 +117,9 @@ public RetrieveResponse retrieve(@NotBlank final String key) throws StorageExcep
       throw t;
     }
   }
+
+  @Override
+  public boolean objectExists(String key) {
+    return (amazonS3.doesObjectExist(bucket, key)) ? true : false;
```
Comment (ahoffer) [inline multi-int-store/src/main/java/com/connexta/multiintstore/adaptors/S3StorageAdaptor.java]: :exclamation:  `doesObjectExist` returns a boolean. No need for the ternary expression.

File: multi-int-store/src/main/java/com/connexta/multiintstore/adaptors/S3StorageAdaptor.java [hunk 2]
```diff
@@ -66,7 +66,7 @@ public void store(
     } catch (RuntimeException | InterruptedException e) {
       throw new StorageException(
           String.format(
-              "Unable to store \"%s\" in bucket \"%s\" with key \"%s\"", fileName, key, bucket),
+              "Unable to store \"%s\" in bucket \"%s\" with key \"%s\"", fileName, bucket, key),
```
Comment (ahoffer) [inline multi-int-store/src/main/java/com/connexta/multiintstore/adaptors/S3StorageAdaptor.java]: Good catch!

File: multi-int-store/src/test/java/com/connexta/multiintstore/SolrTests.java [hunk 3]
```diff
@@ -326,13 +335,63 @@ public String getFilename() {
...
+        IOUtils.toInputStream(metadataContents, metadataEncoding);
+    final long metadataFileSize = (long) metadataInputStream.available();
+    final String metadataFileName = "test_file_name.txt";
+    final String metadataMimeType = "text/plain";
+    final MultiValueMap<String, Object> storeMetadataRequestBody = new LinkedMultiValueMap<>();
+    storeMetadataRequestBody.add("fileSize", String.valueOf(metadataFileSize));
+    storeMetadataRequestBody.add("mimeType", metadataMimeType);
+    storeMetadataRequestBody.add(
+        "file",
+        new InputStreamResource(metadataInputStream) {
```
Comment (ahoffer) [inline multi-int-store/src/test/java/com/connexta/multiintstore/SolrTests.java]: :bulb: We override the contentLength and getFilename methods of InputStreamResource many times in anonymous classes. 

image

Let's create a subclass like:

```
public class KnownInputStreamResource extends InputStreamResource {

    // Pass values into constructor
    public KnownInputStreamResource(InputStream istream, long length, String filename) 


    //Override contentLength() and getFilename()

}
```

 Use this class in place of InputStreamResource

```suggestion
        new KnownInputStreamResource(metadataInputStream, metadataFileSize, metadataFileName);
```

File: multi-int-store/src/test/java/com/connexta/multiintstore/SolrTests.java [hunk 4]
```diff
@@ -326,13 +335,63 @@ public String getFilename() {
...
+            return metadataFileSize;
+          }
+
+          @Override
+          public String getFilename() {
+            return metadataFileName;
+          }
+        });
+    storeMetadataRequestBody.add("fileName", metadataFileName);
+    final HttpHeaders storeMetadataRequestHttpHeaders = new HttpHeaders();
```
Comment (ahoffer) [inline multi-int-store/src/test/java/com/connexta/multiintstore/SolrTests.java]: :bulb: It is probably time for our own HtttpHeaders factory. This might be something for a different PR, at your discretion.

I imagine it like:

```
@Bean
public class HttpHeadersFactory {

@Value("${acceptVersion}")
private final accept-version;

public static final String ACCEPT_VERSION = "Accept-Version"

public HttpHeaders createWithAcceptVersion() {
    headers = new HttpHeaders();
    headers.set(ACCEPT_VERSION,  accept-version);
    return headers;
}

public static HttpHeaders create() {
    return new HttpHeaders().createWithAcceptVersion();
}

```

Then line  381 becomes
```
  new HttpEntity<>(storeMetadataRequestBody, HttpHeadersFactory.create()));
```

We might even want an HttpEntityFactory that uses the HttpHeadersFactory. The client code would not know that an HttpHeadersFactory existed because it is encapsulated by a HttpEntityFactory

File: multi-int-store/src/main/java/com/connexta/multiintstore/adaptors/S3StorageAdaptor.java (hunk 1 above)
Comment (jhunzik) [reply multi-int-store/src/main/java/com/connexta/multiintstore/adaptors/S3StorageAdaptor.java]: Not sure why I didn't see this when I wrote it. 😂

File: multi-int-store/src/test/java/com/connexta/multiintstore/SolrTests.java (hunk 3 above)
Comment (jhunzik) [reply multi-int-store/src/test/java/com/connexta/multiintstore/SolrTests.java]: Good suggestion, I think we should make this a separate PR.

File: multi-int-store/src/test/java/com/connexta/multiintstore/SolrTests.java (hunk 4 above)
Comment (jhunzik) [reply multi-int-store/src/test/java/com/connexta/multiintstore/SolrTests.java]: I definitely think it should be a separate PR. I like the idea, but I'm not sure it's worth doing. Can you foresee a time when we have more than the custom `Accept-Version` header.

**Comment by jhunzik** [comment]
build now

**Comment by jhunzik** [comment]
build now

File: multi-int-store/src/test/java/com/connexta/multiintstore/SolrTests.java [hunk 5]
```diff
@@ -93,16 +95,21 @@ public SolrClientConfiguration testSolrClientConfiguration() {
   private String endpointUrlRetrieve;
 
   @Before
-  public void before() {
+  public void before() throws IOException, SolrServerException {
     when(mockAmazonS3.putObject(any(PutObjectRequest.class)))
         .thenReturn(mock(PutObjectResult.class));
+
+    solrClient.deleteByQuery(SOLR_COLLECTION, "*");
+    solrClient.commit(SOLR_COLLECTION);
```
Comment (emmberk) [inline multi-int-store/src/test/java/com/connexta/multiintstore/SolrTests.java]: We should not need to clear solr between itests. I believe that it is better to write tests that do not have the assumption that Solr is empty, because in a production instance, Solr will never be empty. In each test, we can get the id from the original store request or we can use a fake id to test an id that doesn't exist.

File: multi-int-store/src/test/java/com/connexta/multiintstore/SolrTests.java [hunk 6]
```diff
@@ -326,13 +335,63 @@ public String getFilename() {
   }
 
   @Test
-  @Ignore("TODO verify id matches something in S3 before storing to solr")
-  public void testStoreMetadataProductIdNotFound() {
-    // TODO verify 404? 400?
+  public void testStoreMetadataProductIdNotFound() throws Exception {
+    when(mockAmazonS3.doesObjectExist(anyString(), anyString())).thenReturn(Boolean.TRUE);
```
Comment (emmberk) [inline multi-int-store/src/test/java/com/connexta/multiintstore/SolrTests.java]: We should be able to stub this method only when the id is the specific one that we expect, not any id.

Comment (emmberk) [inline multi-int-store/src/test/java/com/connexta/multiintstore/SolrTests.java]: Why do you stub this method twice?

File: multi-int-store/src/main/java/com/connexta/multiintstore/adaptors/S3StorageAdaptor.java [hunk 7]
```diff
@@ -117,4 +117,9 @@ public RetrieveResponse retrieve(@NotBlank final String key) throws StorageExcep
       throw t;
     }
   }
+
+  @Override
+  public boolean objectExists(String key) {
+    return amazonS3.doesObjectExist(bucket, key);
```
Comment (emmberk) [inline multi-int-store/src/main/java/com/connexta/multiintstore/adaptors/S3StorageAdaptor.java]: Why create this new method instead of throwing an exception in `#store` if the id already exists? We should never support overriding, so `#store` should fail for duplicate id regardless of whether `#objectExists` was checked first.

File: multi-int-store/src/test/java/com/connexta/multiintstore/SolrTests.java (hunk 6 above)
Comment (jhunzik) [reply multi-int-store/src/test/java/com/connexta/multiintstore/SolrTests.java]: That was a mistake. The stub at the beginning of the method isn't necessary. I'll put up a PR to address your comments.
